reactive/etcd.py

```diff
@@ -175,58 +145,190 @@ def install_etcd():
+
+    Before executing self registration, we must adhere to the nature of offline
+    static turnup rules. If we find a GUID in the member list without peering
+    information the unit will enter a race condition and must wait for a clean
+    status output before we can progress to self registration.
+    '''
+    # We're going to communicate with the leader, and we need our bootstrap
+    # startup string once.. TBD after that.
+    etcdctl = EtcdCtl()
+    bag = EtcdDatabag()
+    # Assume a hiccup during registration and attempt a retry
+    if bag.cluster_unit_id:
+        bag.cluster = bag.registration_peer_string
+        render('defaults', '/etc/default/etcd', bag.__dict__)
+        host.service_restart('etcd')
+        time.sleep(2)
+
+    peers = etcdctl.member_list(leader_get('leader_address'))
+    for unit in peers:
+        if 'client_urls' not in peers[unit].keys():
+            # we cannot register. State not attainable.
+            msg = 'Waiting for unit to complete registration'
+            status_set('waiting', msg)
+            return
+
+    if not bag.cluster_unit_id:
+        bag.leader_address = leader_get('leader_address')
+        resp = etcdctl.register(bag.__dict__)
+        if resp and 'cluster_unit_id' in resp.keys() and 'cluster' in resp.keys():
+            bag.cache_registration_detail('cluster_unit_id',
+                                          resp['cluster_unit_id'])
+            bag.cache_registration_detail('registration_peer_string',
+                                          resp['cluster'])
+
+            bag.cluster_unit_id = resp['cluster_unit_id']
+            bag.cluster = resp['cluster']
+
+    render('defaults', '/etc/default/etcd', bag.__dict__)
+    host.service_restart('etcd')
+    time.sleep(2)
+
+    # Check health status before we say we are good
+    etcdctl = EtcdCtl()
+    status = etcdctl.cluster_health()
+    if 'unhealthy' in status:
+        status_set('blocked', 'Cluster not healthy')
+        return
+    open_port(bag.port)
+    set_state('etcd.registered')
 
-    codename = host.lsb_release()['DISTRIB_CODENAME']
-    if codename == 'trusty':
-        templating.render('upstart', '/etc/init/etcd.conf',
-                          cluster_data, owner='root', group='root')
-    else:
-        templating.render('defaults', '/etc/default/etcd',
-                          cluster_data, owner='root', group='root')
 
-    host.service('restart', 'etcd')
-    set_state('etcd.configured')
+@when('etcd.installed')
+@when('etcd.ssl.placed')
+@when('leadership.is_leader')
+@when_not('etcd.leader.configured')
+def initialize_new_leader():
+    bag = EtcdDatabag()
+    bag.token = bag.token
+    bag.cluster_state = 'new'
+    bag.cluster = "{}=https://{}:{}".format(bag.unit_name,
+                                            bag.private_address,
+                                            bag.management_port)
+    render('defaults', '/etc/default/etcd', bag.__dict__, owner='root',
```

Comment:
This is the third time I have encountered the same render, restart, sleep combination in this method. That is always a sign that we could abstract these steps to their own method. 
```python
def render_restart_etcd(file, options, seconds):
    """Render the configuration file with the options and sleep for the provided seconds."""
    render('defaults', file, options, owner='root', group='root')
    host.service_restart('etcd')
    time.sleep(2)
```